pyramid/csrf.py

```diff
@@ -0,0 +1,225 @@
+from functools import partial
+
+from pyramid.compat import (
+    bytes_,
+    urlparse,
+    )
+
+from pyramid.exceptions import (
+    BadCSRFOrigin,
+    BadCSRFToken,
+)
+from pyramid.interfaces import ICSRF
+from pyramid.settings import aslist
+from zope.interface import implementer
+from pyramid.util import (
+    is_same_domain,
+    strings_differ,
+)
+
+
+@implementer(ICSRF)
+class SessionCSRF(object):
+    """ The default CSRF implementation, which mimics the behavior from older
+    versions of Python. The ``new_csrf_token`` and ``get_csrf_token`` methods
+    are indirected to the underlying session implementation.
+
+    .. versionadded :: 1.8a1
+    """
+    def new_csrf_token(self, request):
+        """ Sets a new CSRF token into the session and returns it. """
+        return request.session.new_csrf_token()
+
+    def get_csrf_token(self, request):
+        """ Returns the currently active CSRF token from the session, generating
+        a new one if needed."""
+        return request.session.get_csrf_token()
+
+
+def csrf_token_template_global(event):
+    request = event.get('request', None)
+    try:
+        registry = request.registry
+        csrf = registry.queryUtility(ICSRF)
+        if csrf is not None:
+            event['get_csrf_token'] = partial(csrf.get_csrf_token, request)
+    except AttributeError:
+        pass
+
+
+def get_csrf_token(request):
+    """ Get the currently active CSRF token for the request passed, generating
+    a new one using ``new_csrf_token(request)`` if one does not exist. This
+    calls the equivalent method in the chosen CSRF protection implementation.
+
+    .. versionadded :: 1.8a1
+    """
+    registry = request.registry
+    csrf = registry.queryUtility(ICSRF)
+    if csrf is not None:
+        return csrf.get_csrf_token(request)
```

Comment:
> If third-party code registers a view with require_csrf, AFAIK there is no way straightforward way to disable CSRF protection.

This is intentional... if someone defines a view that explicitly requires csrf then it will be inaccessible until someone registers a csrf policy for the app. Turning off csrf protection for those views is a non-goal and can already be done just by re-defining it after `config.commit()` to override the old view.

I'd expect this code to raise if there is no policy. Simply using `getUtility` should be enough.